internal/configs/module_deprecations.go

```diff
@@ -0,0 +1,144 @@
+}
+
+type ModuleDeprecationDiagnosticExtraDeprecationItem struct {
+	Version            string `json:"version"`
+	SourceName         string `json:"source_name"`
+	DeprecationMessage string `json:"deprecation_message"`
+	Link               string `json:"link"`
+}
+
+func (i *WorkspaceDeprecationInfo) HasDeprecations() bool {
+	if i == nil || i.ModuleDeprecationInfos == nil {
+		return false
+	}
+	for _, deprecationInfo := range i.ModuleDeprecationInfos {
+		if deprecationInfo != nil && deprecationInfo.hasDeprecations() {
+			return true
+		}
+	}
+	return false
+}
+
+func (i *ModuleDeprecationInfo) hasDeprecations() bool {
+	if i.RegistryDeprecation != nil {
+		return true
+	}
+	for _, dependencyDeprecationInfo := range i.ExternalDependencies {
+		if dependencyDeprecationInfo != nil && dependencyDeprecationInfo.hasDeprecations() {
+			return true
+		}
+	}
+	return false
+}
+
+// Deprecation info is placed as an string in the Diagnostic Detail for console view,
+// as well as placed in the Diagnostic Extra for parsing for the SRO view in HCP Terraform
+func (i *WorkspaceDeprecationInfo) BuildDeprecationWarning() *hcl.Diagnostic {
+	modDeprecationStrings := []string{}
+	var deprecationList []*ModuleDeprecationDiagnosticExtraDeprecationItem
+	for _, modDeprecationInfo := range i.ModuleDeprecationInfos {
+		if modDeprecationInfo != nil && modDeprecationInfo.RegistryDeprecation != nil {
+			modDeprecation := fmt.Sprintf("\x1b[1mVersion %s of %s\x1b[0m", modDeprecationInfo.RegistryDeprecation.Version, modDeprecationInfo.SourceName)
+			// Link and Message are optional fields, if unset they are an empty string by default
+			if modDeprecationInfo.RegistryDeprecation.Message != "" {
+				modDeprecation = modDeprecation + fmt.Sprintf("\n\n%s", modDeprecationInfo.RegistryDeprecation.Message)
+			}
+			if modDeprecationInfo.RegistryDeprecation.Link != "" {
+				modDeprecation = modDeprecation + fmt.Sprintf("\n\nLink for more information: %s", modDeprecationInfo.RegistryDeprecation.Link)
+			}
+			deprecationList = append(deprecationList, &ModuleDeprecationDiagnosticExtraDeprecationItem{
+				Version:            modDeprecationInfo.RegistryDeprecation.Version,
+				SourceName:         modDeprecationInfo.SourceName,
+				DeprecationMessage: modDeprecationInfo.RegistryDeprecation.Message,
+				Link:               modDeprecationInfo.RegistryDeprecation.Link,
+			})
+			modDeprecationStrings = append(modDeprecationStrings, modDeprecation)
+		}
+		deprecationStrings, deprecationStructs := buildChildModuleDeprecations(modDeprecationInfo.ExternalDependencies, []string{modDeprecationInfo.SourceName})
+		deprecationList = append(deprecationList, deprecationStructs...)
+		modDeprecationStrings = append(modDeprecationStrings, deprecationStrings...)
+	}
+	deprecationsMessage := ""
+	for _, deprecationString := range modDeprecationStrings {
+		deprecationsMessage += deprecationString + "\n\n"
+	}
+
+	return &hcl.Diagnostic{
```

Comment:
I've got a modest architectural suggestion for you here — instead of re-using the hcl.Diagnostic type here, what do you think about making your own DeprecationsDiagnostic struct, shoving the flattened deprecationList (and a final boolean hasDeprecations result) into it, and [implementing `tfdiags.Diagnostic` on it](https://github.com/hashicorp/terraform/blob/a79e38cf3a3bb7611306aefcd2361df68d276493/internal/tfdiags/diagnostic.go#L12-L29)? 

I think it could yield some cleaner-feeling code, but it's not something I'd necessarily insist on.

Reply:
I do quite like the idea! My only worry is the amount of rework to get it there– or at least to make it so it's not more confusing 😅.

Reply:
I did play around with it and completely agree with the idea, just worried it'll add to overhead in reviewing since it'll be a big change from the current implementation. Absolutely noted for the future though 😁